File: DataStrucutres - Python/DoubleLinkedList/double_linked_list.py
class Link:
  def __init__(self, data):
    self.data = data
    self.next = None
    self.prev = None

class LinkIterator:
  def __init__(self, link):
    self.next_link = link
    self.count = 0

  def next(self):
    link = self.next_link
    if self.count == 0:
      self.count += 1
    else:
      link = self.next_link = self.next_link.next
    if self.next_link is None:
      raise StopIteration()
    return self.next_link


class DoubleLinkedList:
  def __init__(self):
    self.first = None
    self.last = None
    self.size = 0

  def empty(self):
    return (self.size == 0)

  def append(self, data):
    link = Link(data)
    if self.empty():
      self.first = self.last = link
    else:
      link.prev = self.last
      self.last.next = link
      self.last = link
    self.size += 1
    return link

  def prepend(self, data):
    link = Link(data)
    if self.empty():
      self.first = self.last = link
    else:
      link.next = self.first
      self.first.prev = link
      self.first = link
    self.size += 1
    return link

  def insertAfter(self, data, link):
    new_link = Link(data)
    if link.next is not None:
      new_link.next = link.next
      link.next.prev = new_link
    new_link.prev = link
    link.next = new_link
    return new_link

  def find(self, data):
    for link in self:
      if data == link.data:
        return link
    return None

  def delete(self, link):
    if link is not None:
      if link.prev is None: # Is the first one
        next_link = self.first.next
        if next_link is not None:
          next_link.prev = None
        else:
          self.last = None
        self.first = next_link
      elif link.next is None: # Is the last one
        last_link = self.last.prev
        if last_link is not None:
          last_link.next = None
        else:
          self.first = None
        self.last = last_link
      else: # In the middle
        link.prev.next = link.next
        link.next.prev = link.prev
      self.size -= 1
      return link
    return None

  def __iter__(self):
    return LinkIterator(self.first)

  def __contains__(self, data):
    return self.find(data) != None

File: DataStructures/implementations/src/main/java/datastructures/DoubleLinkedList.java
package datastructures;

import java.util.Iterator;

public class DoubleLinkedList<T> implements Iterable<DoubleLinkedList.Link> {
  private Link<T> first;
  private Link<T> last;
  private int size = 0;

  public Link<T> append(T data) {
    Link<T> link = new Link<>(data);
    if (isEmpty()) {
      first = last = link;
    } else {
      last.next = link;
      link.prev = last;
      last = link;
    }
    size++;
    return link;
  }

  public Link<T> prepend(T data) {
    Link<T> link = new Link<>(data);
    if (isEmpty()) {
      first = last = link;
    } else {
      first.prev = link;
      link.next = first;
      first = link;
    }
    size++;
    return link;
  }

  public Link<T> find(T data) {
    for (Link<T> link : this) {
      if (link.data.equals(data)) {
        return link;
      }
    }
    return null;
  }

  public Link<T> delete(Link<T> linkToDelete) {
    if (linkToDelete == null) return null;
    if (linkToDelete.prev == null) deleteFirst();
    else if (linkToDelete.next == null) deleteLast();
    else {
      linkToDelete.next.prev = linkToDelete.prev;
      linkToDelete.prev.next = linkToDelete.next;
    }

    size--;
    return linkToDelete;
  }

  public Link<T> deleteFirst() {
    Link<T> temp = this.first;
    first.next.prev = null;
    first = first.next;
    return temp;
  }

  public Link<T> deleteLast() {
    Link<T> temp = last;
    last.prev.next = null;
    last = last.prev;
    return temp;
  }

  public Link<T> getFirst() {
    return first;
  }

  public Link<T> getLast() {
    return last;
  }

  public boolean isEmpty() {
    return size == 0;
  }

  public Link<T> insertAfter(Link<T> link, T data) {
    Link<T> newLink = new Link<>(data);
    if (link.next == null) newLink = append(data);
    else {
      link.next.prev = newLink;
      newLink.next = link.next;
      newLink.prev = link;
      link.next = newLink;
      size++;
    }
    return newLink;
  }

  public Link<T> inserBefore(Link<T> link, T data) {
    Link<T> newLink = new Link<>(data);
    if (link.prev == null) newLink = prepend(data);
    else {
      link.prev.next = newLink;
      newLink.prev = link.prev;
      newLink.next = link;
      link.prev = newLink;
      size++;
    }
    return newLink;
  }

  @Override
  public Iterator<Link> iterator() {
    return new LinkIterator(first, false);
  }

  public Iterator<Link> reversedIterator() {
    return new LinkIterator(last, true);
  }


  class LinkIterator implements Iterator<Link> {
    private Link<T> current;
    private boolean reversed;

    public LinkIterator(Link<T> link, boolean reversed) {
      current = link;
      this.reversed = reversed;
    }

    @Override
    public boolean hasNext() {
      return (current != null);
    }

    @Override
    public Link<T> next() {
      Link<T> temp = current;
      current = nextLink(current);
      return temp;
    }


    private Link<T> nextLink(Link<T> link) {
      if (reversed) return link.prev;
      return link.next;
    }
  }

  public class Link<T> {
    T data;
    Link<T> next;
    Link<T> prev;

    public Link(T data) {
      this.data = data;
    }
  }
}